### PR TITLE
mgr/dashboard: add api for rgw multisite replication wizard

### PR DESCRIPTION
**Pull request by aaSharma14**
Fixes: https://tracker.ceph.com/issues/66227

https://github.com/ceph/ceph/assets/66050535/675b67c3-b9f9-4d3c-9242-ae716d9eb5ee

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 1]
```diff
@@ -111,6 +111,27 @@ def migrate(self, daemon_name=None, realm_name=None, zonegroup_name=None, zone_n
                                                          secret_key)
         return result
 
+    @RESTController.Collection(method='PUT', path='/setup_multisite_replication')
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/rgw.py:114]: this sounds like a POST action to me and we can simply call the endpoint `POST /multisite-replications` rather than `setup_multisite_replication`. And as per the REST API standards, its better to use hyphens to separate the words instead of underscores, which will improve readability and easability.  
```suggestion
    @RESTController.Collection(method='PUT', path='/setup_multisite_replication')
```

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 2]
```diff
@@ -111,6 +111,27 @@ def migrate(self, daemon_name=None, realm_name=None, zonegroup_name=None, zone_n
...
+                                    zonegroup_endpoints=None, zone_name=None, zone_endpoints=None,
+                                    username=None, cluster=None):
+        multisite_instance = RgwMultisite()
+        result = multisite_instance.setup_multisite_replication(realm_name, zonegroup_name,
+                                                                zonegroup_endpoints, zone_name,
+                                                                zone_endpoints, username,
+                                                                cluster)
+        return result
+
+    @RESTController.Collection(method='GET', path='/restart_rgw_daemons_and_set_credentials')
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/rgw.py:127]: same here. its not really good to call an endpoint `restart_rgw_daemons_and_set_credentials`. We should name it short and clear. and why is this a `GET` endpoint. It is doing some action so it should be different, maybe PUT?

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 3]
```diff
@@ -111,6 +111,27 @@ def migrate(self, daemon_name=None, realm_name=None, zonegroup_name=None, zone_n
                                                          secret_key)
         return result
 
+    @RESTController.Collection(method='PUT', path='/setup_multisite_replication')
+    @allow_empty_body
+    # pylint: disable=W0102,W0613
+    def setup_multisite_replication(self, daemon_name=None, realm_name=None, zonegroup_name=None,
+                                    zonegroup_endpoints=None, zone_name=None, zone_endpoints=None,
+                                    username=None, cluster=None):
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/rgw.py:119]: By cluster you mean fsid of the cluster?
```suggestion
    def setup_multisite_replication(self, daemon_name=None, realm_name=None, zonegroup_name=None,
                                    zonegroup_endpoints=None, zone_name=None, zone_endpoints=None,
                                    username=None, cluster_fsid=None):
```

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 4]
```diff
@@ -1061,6 +1066,115 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+        def get_updated_endpoints(endpoints):
+            # Update endpoint URLs by replacing hostnames with IP addresses.
+            hostname_to_ip = {host['hostname']: host['addr'] for host in (h.to_json() for h in orch.hosts.list())}
+            return [self.replace_hostname(endpoint, hostname_to_ip) for endpoint in endpoints.split(',')]
+
+        zonegroup_ip_url = ','.join(get_updated_endpoints(zonegroup_endpoints))
+        zone_ip_url = ','.join(get_updated_endpoints(zone_endpoints))
+
+        # Create the realm and zonegroup
+        self.create_realm(realm_name, True)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1092]: for readability, can you use keyword arguments rather than using positional arguments here and below codes? It makes it easier for the reviewers to see what argument is assigned with what value.
```suggestion
        self.create_realm(realm_name, default=True)
```

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 5]
```diff
@@ -1061,6 +1066,115 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+        self.create_zonegroup(realm_name, zonegroup_name, True, True, zonegroup_ip_url)
+
+        # Create the zone and system user, then modify the zone with user credentials
+        if self.create_zone(zone_name, zonegroup_name, True, True, zone_ip_url, None, None):
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1096]: same here

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 6]
```diff
@@ -1061,6 +1066,115 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+            if user_details:
+                keys = user_details['keys'][0]
+                self.modify_zone(zone_name, zonegroup_name, True, True, zone_ip_url, keys['access_key'], keys['secret_key'])
+
+        # Restart RGW daemons and set credentials
+        self.restart_rgw_daemons_and_set_credentials()
+
+        # Get realm tokens and import to another cluster if specified
+        realm_token_info = CephService.get_realm_tokens()
+        time.sleep(5)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1107]: i don't like the idea of explicitly adding wait time. why do we need this? how can we eliminate this?

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 7]
```diff
@@ -1061,6 +1066,115 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+        if cluster_token:
+            headers = {
+                'Accept': 'application/vnd.ceph.api.v1.0+json',
+                'Authorization': 'Bearer ' + cluster_token,
+            }
+        else:
+            headers = {
+                'Accept': 'application/vnd.ceph.api.v1.0+json',
+                'Content-Type': 'application/json',
+            }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1135]: if there is no token, what kind of request are we sending?

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 8]
```diff
@@ -1061,6 +1066,115 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+            'placement_spec': placement_spec
+        }
+
+        if not cluster_url.endswith('/'):
+            cluster_url = cluster_url + '/'
+
+        path = 'api/rgw/realm/import_realm_token'
+
+        try:
+            response = requests.request('POST', cluster_url + path, json=payload, headers=headers, verify=False)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1154]: never put verify=False. The default should be True and the certificates should be passed over to the API to ensure the security.
```suggestion
            response = requests.request('POST', cluster_url + path, json=payload, headers=headers, verify=True)
```

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 9]
```diff
@@ -1061,6 +1066,115 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+            response.raise_for_status()
+            if response.status_code == 200:
+                restart_path = 'ui-api/rgw/multisite/restart_rgw_daemons_and_set_credentials'
+                restart_gateways_response = requests.request('GET', cluster_url + restart_path, headers=headers, verify=False)
+                restart_gateways_response.raise_for_status()
+                if restart_gateways_response.status_code == 200:
+                    return True
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1161]: IMO, we should be reusing the multi-cluster codes to do these to avoid duplicates. can you try to reuse it from the multi-cluster code?

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 10]
```diff
@@ -1061,6 +1066,115 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+            raise DashboardException(f"Error parsing Dashboard API response: {e.msg}", component='dashboard')
+
+    def restart_rgw_daemons_and_set_credentials(self):
+        # Restart RGW daemons and set credentials.
+        orch = OrchClient.instance()
+        daemons = [d.to_dict() for d in orch.services.list_daemons()]
+        daemon_list = [d for d in daemons if d['daemon_type'] == 'rgw']
+        for daemon in daemon_list:
+            orch.daemons.action(daemon['daemon_name'], 'restart', None)
+        time.sleep(5)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1175]: dont do this. if what you want is to wait for the daemon to be in `running ` state, explicitly check for the status of the daemon to be in running state and then proceed to configure rgw credentials. Because now you are assuming rgw daemon to be up after 5 sec but are you sure it won't go into an error state, or unknown state after 5 seconds?

**Review by nizamial09 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 6 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1110]: what about this?

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 10 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1173]: what about this?

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 5 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1097]: "for readability, can you use keyword arguments rather than using positional arguments here and below codes? It makes it easier for the reviewers to see what argument is assigned with what value."

something I said in the above comment.

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 6 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/services/rgw_client.py]: Removed this one, wasn't needed anymore

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 10 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1224]: I have added an additional check after time.sleep to check if daemons are in a running state and then proceed to configure the credentials. But time.sleep is still required because if after we restart the daemons, the status of the daemons doesn't change and we need to wait for a few seconds to again check if the status has changed or not.

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 11]
```diff
@@ -1112,6 +1118,127 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+                                         component='dashboard')
+
+    def restart_rgw_daemons_and_set_credentials(self):
+        # Restart RGW daemons and set credentials.
+        orch = OrchClient.instance()
+        daemons = [d.to_dict() for d in orch.services.list_daemons()]
+        daemon_list = [d for d in daemons if d['daemon_type'] == 'rgw']
+        for daemon in daemon_list:
+            orch.daemons.action(daemon['daemon_name'], 'restart', None)
+        time.sleep(5)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1224]: I was looking into the cephadm interfaces and saw that we already have a way to restart a service by utilizing the [service_action](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/orchestrator/_interface.py#L626) interface. Then I checked the dashboard code and I saw we already have that one [exposed for us](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/services/orchestrator.py#L128).

Let's use that one rather than iterate through the list of daemons and restarting the one by one.

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 12]
```diff
@@ -1112,6 +1118,127 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+        orch = OrchClient.instance()
+        start_time = time.time()
+        while time.time() - start_time < timeout:
+            daemons = [d.to_dict() for d in orch.services.list_daemons()]
+            restarted_daemons = [d for d in daemons if d['daemon_type'] == 'rgw' and d['status'] == 1]
+            if all(daemon in restarted_daemons for daemon in daemon_list):
+                logger.info("All RGW daemons restarted successfully")
+                return True
+            time.sleep(interval)
+        raise TimeoutError("RGW daemons did not restart within the timeout period.")
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1240]: once you restart a service/daemon, if you want to know if the service/daemon is restarted you can rely on the `last_refresh` status of that service. Because every action will trigger the cephadm serve and later on can change the status of the daemon.. So rather than relying on more hard-coded wait (which is unreliable and can cause errors and performance issues), let's rely on actual conditions and adapt the logic accordingly. 

Below is a logic I wrote locally on how I would vision this check
```
def verify_service_restart(self, type: str, id: str):
        orch = OrchClient.instance()
        service_name = f'{type}.{id}'
        info = orch.services.get(service_name)[0].to_dict()
        last_refreshed = info['status']['last_refresh']
        orch.services.reload(type, id)

        while True:
            updated_info = orch.services.get(service_name)[0].to_dict()
            if updated_info['status']['last_refresh'] != last_refreshed:
                break

        import time
        start_time = time.time()
        while True:
            daemon = [d.to_dict() for d in orch.services.list_daemons(service_name=service_name)]
            if daemon[0]['status']['state'] == 'starting':
                continue
            if daemon[0]['status']['state'] == 'running':
                return True
            if daemon[0]['status']['state'] in ('unknown', 'error', 'stopped'):
                raise DashboardException(
                    code='daemon_restart_failed',
                    msg=f"Failed to restart the daemon {service_name}.
                    Daemon state is {daemon[0]['status']['state']}."
                )
            if time.time() - start_time > 10:
                raise DashboardException(
                    code='daemon_restart_timeout',
                    msg=f"Timeout reached while waiting for daemon {service_name} to start."
                )
```

and this particular code block doesn't belong to rgw_client.py. Rather than growing this file with unrelated codes, move this to another file. We don't have a `service.py` services file so let's create a new service.py file and put it there and reuse this wherever we want.


Additionally I saw the orch.services.reload() is broken since its using an outdated code. Let's update that too to something like below
```

    @wait_api_result
    def reload(self, service_type, service_ids) -> OrchResult[List[str]]:
        if not isinstance(service_ids, list):
            service_ids = [service_ids]

        for service_id in service_ids:
            self.api.service_action('restart', f'{service_type}.{service_id}')
```

I haven't done proper testing to the above so please make sure you are testing these functions properly.

File: src/pybind/mgr/mgr_module.py [hunk 13]
```diff
@@ -2445,7 +2445,7 @@ def send_rgwadmin_command(self, args: List[str],
                 cmd,
                 stdout=subprocess.PIPE,
                 stderr=subprocess.PIPE,
-                timeout=10,
+                timeout=30,   # changed for testing purpose, will be removed once ready to merge
```
Comment (nizamial09) [inline src/pybind/mgr/mgr_module.py:2448]: don;t forget to remove this

**Review by nizamial09 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 12 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1240]: also please improve the code block accordingly based on your testing. I still see there are a couple of rooms for improvements

File: src/pybind/mgr/mgr_module.py (hunk 13 above)
Comment (aaSharma14) [reply src/pybind/mgr/mgr_module.py:2448]: will do, thanks

**Comment by nizamial09** [comment]
@aaSharma14 everything looks good to me, thank you. One thing


> and this particular code block doesn't belong to rgw_client.py. Rather than growing this file with unrelated codes, move this to another file. We don't have a service.py services file so let's create a new service.py file and put it there and reuse this wherever we want.

can you please take this as well. We shouldn't grow our rgw_client code with codes that shouldn't belong in them. As it is right now, the file has grown more than 1000s of line. thanks

File: src/pybind/mgr/dashboard/services/service.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+            if all_running:
+                return True
+
+            if time.time() - start_time > 10:
+                raise DashboardException(
+                    code='daemon_restart_timeout',
+                    msg=f"Timeout reached while waiting for daemon {service_name} to start."
+                )
+            return False
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/service.py:93]: these are generic to every service and not just rgw so it would be helpful in the future if this is moved to a more generic ServiceManager or as a standalone function outside this class..

**Review by nizamial09 — APPROVED** [review]
lgtm, thanks for addressing my comments..

File: src/pybind/mgr/dashboard/services/service.py (hunk 14 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/services/service.py]: Done. Thanks

**Comment by aaSharma14** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py [hunk 15]
```diff
@@ -230,7 +231,7 @@ def load_multi_cluster_config(self):
 
     @Endpoint('PUT')
     @UpdatePermission
-    def set_config(self, config: object):
+    def set_config(self, config: Dict[str, Any]):
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/multi_cluster.py:234]: do we need to do this here? it looks unrelated. I don't think this will break anything but since we don't have an e2e for multi-cluster yet, its better to skip this to a separate commit or a PR of its own..

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 16]
```diff
@@ -1142,6 +1069,177 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+                    logger.error("Error parsing Dashboard API response: %s", e.msg)
+                    raise DashboardException(f"Error parsing Dashboard API response: {e.msg}",
+                                             component='dashboard')
+        except Exception as e:
+            logger.error("Failed to import realm token to cluster: %s", e)
+            raise
+
+    def check_user_in_second_cluster(self, cluster_url, cluster_token, username):
+        logger.info("Checking for user %s in the second cluster", username)
+        path = 'api/rgw/zone/get_user_list?zoneName=new_replicated_zone'
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1212]: where is this `new_replicated_zone` coming from? is that hard-coded or does user has a say?

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 17]
```diff
@@ -1142,6 +1069,177 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+        user_found = False
+        start_time = time.time()
+        while not user_found:
+            if time.time() - start_time > 300:  # Timeout after 5 minutes
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1216]: 5 mins? does it take that longer.. and should worst case be that longer. 5mins for an api is unacceptable from what i read. Usually it should be around 30s to 2min. won't we get the response under it?

if its just a local issue, try to increase your kcli resource and try again and see if it has any improvements but blocking api for 5 mins won't be nice. 
@cloudbehl ^

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 18]
```diff
@@ -1142,6 +1069,177 @@ def migrate_to_multisite(self, realm_name: str, zonegroup_name: str, zone_name:
...
+                        # pylint: disable=protected-access
+                        restart_daemons_content = multi_cluster_instance._proxy(method='PUT', base_url=cluster_url,  # noqa E501  # pylint: disable=line-too-long
+                                                                                path='ui-api/rgw/multisite/setup-rgw-credentials',  # noqa E501  # pylint: disable=line-too-long
+                                                                                token=cluster_token)  # noqa E501  # pylint: disable=line-too-long
+                        logger.info("Restarted RGW daemons in the second cluster: %s", restart_daemons_content)  # noqa E501  # pylint: disable=line-too-long
+                        break
+            except requests.RequestException as e:
+                logger.error("Error checking user in the second cluster: %s", e)
+            logger.info("User %s not found yet, retrying in 10 seconds", username)
+            time.sleep(10)
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/services/rgw_client.py:1241]: ?
```suggestion
            time.sleep(5)
```

File: src/pybind/mgr/mgr_module.py (hunk 13 above)
Comment (nizamial09) [reply src/pybind/mgr/mgr_module.py:2448]: ?

File: src/pybind/mgr/dashboard/controllers/multi_cluster.py (hunk 15 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/controllers/multi_cluster.py:234]: @nizamial09 , I was getting mypy error related to this in make-check so had to make this change. Am not sure if a separate PR just for this change makes sense.

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 16 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1212]: This is the new zone that is going to be created in the secondary cluster. There's no user input for this..so yes it is hard coded

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 17 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1216]: This is the timeout for the worst case scenario that I encountered in my local env. Its not like it will take 5 mins every time. We can anyways modify this when we have inputs from QE team and they test it on real env.

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 17 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/services/rgw_client.py]: Updated the PR with a timeout of 2 mins

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/mgr_module.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/mgr_module.py]: Already removed
